Fix specifications in markdown documentation

shym opened the pull request:
Add a few `Makefile` rules to check the Gospel specifications inserted as OCaml code blocks by extracting the code blocks to a corresponding *.mli file and calling `gospel check` on it
Skip the documentation files intentionally using invalid syntax or citing code blocks that cannot all be exported to one file
Add the generated *.mli files to .gitignore

This reveals issues in quite a few documentation files

shym commented:
I’m not sure how to fix the syntax errors on `ephemeral`.

pascutto commented on Makefile:
```diff
@@ -7,10 +7,38 @@ install:
+
+# To check that the markdown documentation is kept valid
+
+GOSPEL_CMD := dune exec --display=quiet -- bin/cli.exe
```
```suggestion
GOSPEL_CMD := dune exec --display=quiet -- gospel
```

pascutto commented on Makefile:
````diff
@@ -7,10 +7,38 @@ install:
+.PRECIOUS: $(DOCUMENTS)
+
+%.mli: %.md
+	awk '/^ *```ocaml/,/^ *```$$/ { if ($$0 !~ "^ *```") print }' < $< > $@
````
I'm really bad at this so I won't try to fix it, but
```
val f : t
(*@ ... *)

val g : t
(*@ ... *)
```
seems to be rewritten as 
```
val f : t

(*@ ... *)
val g : t

(*@ ... *)
```
which could (should) be an incorrect attachment of the specs.

pascutto reviewed (commented):
Thanks!
Do you think you could make this happen as part of the docs CI?
Also, is there a reason to do it using a `Makefile` rather than a dune rule plugged to the tests? That would also avoid polluting the workspace with the temporary test files.

About the `ephemeral` issue, this is a bug in the preprocessor, not attaching the spec attribute to the string in the floating `[@@@gospel "type t"]` attribute because of the interleaved comment.
The preprocessor doesn't know (yet) about comments or docstrings, so it just treats the comment as a string it shouldn't touch (usually OCaml code) and tries to attach the spec to it, which is indeed a syntax error.

shym commented:
Quasi-complete re-roll of that PR:
- using `dune` rules instead of Makefile (as a side-note, I find it a lot more inefficient (and verbose!) when fixing the syntax errors, because errors are reported against the `.mli` file that is hidden away somewhere, I cannot use standard tooling to go to the location and see the error in more context than what `gospel check` reports)
- generate .mli with skipping lines, to avoid ambiguities,
- fixing a large part of the syntax errors and annotating code blocks that are not using valid syntax.
This triggers 2 `gospel` crashes that I didn’t investigate.

shym commented:
A large update of that PR to prepare its merge for the upcoming release:

- rebased on top of #312, to store also in `test/utils` the code that tests the code extracts in markdown documentation (without creating unduly conflicts later), so that this PR is really only the last 4 commits,
- used the location information so that syntax errors are reported to the markdown file when possible (`gospel pps` breaks that, maybe it could pick up `# line` data to update its location accordingly?),
- fixed up a couple of other syntax errors I found,
- commented on the code extracts that are tagged as _invalid_ with the reason why, to help fix those discrepancies between the documentation and `gospel check` (in particular on `ephemeral`).

shym commented:
Rebased on `main`, now that #312 has been merged.

n-osborne reviewed (approved):
Nice automation that already proven its worth by all the correction!
Thanks!

shym commented:
Approved by human & CI, and it is used as the starting point of another PR, so I merge it now.
